File: src/main/java/ru/umeta/libraryintegration/fs/StringHashFsPersister.java
package ru.umeta.libraryintegration.fs;

import org.apache.commons.codec.DecoderException;
import org.apache.commons.io.FileUtils;
import org.apache.commons.io.LineIterator;
import org.apache.commons.io.output.FileWriterWithEncoding;
import org.apache.commons.codec.binary.Hex;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import ru.umeta.libraryintegration.model.StringHash;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Created by k.kosolapov on 12/2/2015.
 */
@Component
public class StringHashFsPersister {

    private static final String SEPARATOR = "|";
    private static final String UTF_8 = "UTF-8";
    private final ExecutorService executorService = Executors.newSingleThreadExecutor();

    private final BlockingQueue<StringHash> queue = new LinkedBlockingQueue<>();

    private final File storageFile = new File("stringHash.blob");



    public StringHashFsPersister() {
        //storageFile.mkdirs();
        if (!storageFile.exists()) {
            try {
                storageFile.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        executorService.execute(() -> {
            try (FileWriterWithEncoding writerWithEncoding = new FileWriterWithEncoding(storageFile, Charset.forName(UTF_8), true)) {
                while (true) {
                    StringHash stringHash;
                    try {
                        stringHash = queue.take();
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                        return;
                    }
                    writerWithEncoding.write(Hex.encodeHexString(new byte[] {
                            stringHash.getHashPart1(),
                            stringHash.getHashPart2(),
                            stringHash.getHashPart3(),
                            stringHash.getHashPart4()
                    }));
                    writerWithEncoding.write(SEPARATOR);
                    writerWithEncoding.write(String.valueOf(stringHash.getId()));
                    writerWithEncoding.write(SEPARATOR);
                    writerWithEncoding.write(stringHash.getValue());
                    writerWithEncoding.write(SEPARATOR + "\n");
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    public void save(StringHash stringHash) {
        queue.add(stringHash);
    }

    public long fillMaps(Map<String, StringHash> map, Map<Long, StringHash> idMap) {
        long lastId = 0;
        try {
            LineIterator it = FileUtils.lineIterator(storageFile, UTF_8);
            try {
                while (it.hasNext()) {
                    try {
                        StringHash stringHash = new StringHash();
                        String line = it.nextLine();
                        String[] splitStrings = StringUtils.tokenizeToStringArray(line, SEPARATOR);

                        byte[] bytes = Hex.decodeHex(splitStrings[0].toCharArray());
                        if (bytes.length < 4) {
                            continue;
                        } else {
                            stringHash.setHashPart1(bytes[0]);
                            stringHash.setHashPart2(bytes[1]);
                            stringHash.setHashPart3(bytes[2]);
                            stringHash.setHashPart4(bytes[3]);
                        }
                        long id = Long.parseLong(splitStrings[1]);
                        stringHash.setId(id);
                        lastId = Math.max(id, lastId);
                        String value = splitStrings[2];
                        stringHash.setValue(value);

                        map.put(value, stringHash);
                        idMap.put(id, stringHash);

                    } catch (DecoderException e) {
                        e.printStackTrace();
                    }
                }

                }
            finally {
                LineIterator.closeQuietly(it);
            }
        } catch (IOException e) {
            System.out.println("Cannot open the storage file");
        }
        return lastId;
    }
}

File: src/main/java/ru/umeta/libraryintegration/fs/StringHashFsPersister.kt
package ru.umeta.libraryintegration.fs

import gnu.trove.map.hash.TIntLongHashMap
import gnu.trove.map.hash.TLongIntHashMap
import gnu.trove.set.hash.TIntHashSet
import org.apache.commons.codec.DecoderException
import org.apache.commons.codec.binary.Hex
import org.apache.commons.io.FileUtils
import org.apache.commons.io.LineIterator
import org.apache.commons.io.output.FileWriterWithEncoding
import org.springframework.util.StringUtils
import ru.umeta.libraryintegration.model.StringHash
import ru.umeta.libraryintegration.service.getTokens
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.util.*
import java.util.concurrent.Executors

/**
 * Created by k.kosolapov on 12/2/2015.
 */
object StringHashFsPersister : AutoCloseable {

    const val SEPARATOR = "|"
    const val UTF_8 = "UTF-8"

    private val executorService = Executors.newSingleThreadExecutor()

    private val storageFile = File("stringHash.blob")

    private val writerWithEncoding: FileWriterWithEncoding

    init {
        if (!storageFile.exists()) {
            try {
                storageFile.createNewFile()
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        writerWithEncoding = FileWriterWithEncoding(storageFile, Charset.forName(UTF_8), true)
    }

    fun save(stringHash: StringHash, value: String) {
        executorService.execute {
            try {
                writerWithEncoding.write(Hex.encodeHexString(
                        byteArrayOf(stringHash.hashPart1(),
                                stringHash.hashPart2(),
                                stringHash.hashPart3(),
                                stringHash.hashPart4())))
                writerWithEncoding.write(SEPARATOR)
                writerWithEncoding.write(stringHash.id.toString())
                writerWithEncoding.write(SEPARATOR)
                writerWithEncoding.write(value)
                writerWithEncoding.write(SEPARATOR + "\n")
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    fun fillMaps(mapHashCodeToId: TIntLongHashMap, mapIdToSimHash: TLongIntHashMap,
                 mapIdToTokens: HashMap<Long, String>): Long {
        var lastId: Long = 0
        try {
            val it = FileUtils.lineIterator(storageFile, UTF_8)
            try {
                while (it.hasNext()) {
                    try {

                        val line = it.nextLine()
                        var splitStrings = StringUtils.tokenizeToStringArray(line, SEPARATOR)

                        if (splitStrings.size != 3) {
                            if (splitStrings.size == 2 && "00000000" == splitStrings[0]) {
                                val newSplitStrings = arrayOfNulls<String>(3)
                                newSplitStrings[0] = splitStrings[0]
                                newSplitStrings[1] = splitStrings[1]
                                newSplitStrings[2] = ""
                                splitStrings = newSplitStrings
                            } else {
                                continue
                            }
                        }

                        val bytes = Hex.decodeHex(splitStrings[0].toCharArray())
                        val hashPart1: Byte
                        val hashPart2: Byte
                        val hashPart3: Byte
                        val hashPart4: Byte

                        if (bytes.size < 4) {
                            continue
                        } else {
                            hashPart1 = bytes[0]
                            hashPart2 = bytes[1]
                            hashPart3 = bytes[2]
                            hashPart4 = bytes[3]
                        }
                        val id = splitStrings[1].toLong()
                        lastId = Math.max(id, lastId)
                        val value = splitStrings[2]
                        //val tokens = getTokens(value)
                        val simHash = StringHash.Util.collectParts(hashPart1, hashPart2, hashPart3, hashPart4)

                        mapHashCodeToId.put(value.hashCode(), id)
                        mapIdToSimHash.put(id, simHash)
                        mapIdToTokens.put(id, value)

                    } catch (e: DecoderException) {
                        e.printStackTrace()
                    }

                }

            } finally {
                LineIterator.closeQuietly(it)
            }
        } catch (e: IOException) {
            println("Cannot open the storage file")
        }

        return lastId
    }

    override fun close() {
        writerWithEncoding.close()
    }
}
